RA-07-categories-screen

DenisGubernatorov opened the pull request:
1) Добавление инастройка компонентов для экрана "Категории"

DenisGubernatorov committed:
RA-07-categories-screen
1) В fragment_list_categories.xml добавлен header с изображением и текстом.
2) Настройка добавленных элементов

DenisGubernatorov committed:
RA-07-categories-screen
1) Добавлен RecyclerView и layout для него.
2) Настройка атрибутов для новых компонентов

DenisGubernatorov committed:
RA-07-categories-screen
1) RecyclerView добавлен вертикальный отступ от шапки

aGroupoid commented on app/src/main/res/layout/item_category.xml:
```diff
@@ -0,0 +1,16 @@
+        android:foreground="?attr/selectableItemBackground"
+        app:cardCornerRadius="@dimen/cat_corner_radius_header"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent" />
```
На макете фрагмента список карточек должен выглядеть также как в фигме. Карточку нужно сверстать до конца и заполнить тестовыми данными через tools.
Также нужно учесть ширину экрана, чтоб карточки занимали всё место по ширине вне зависимости от размеров экрана

aGroupoid commented on app/src/main/res/layout/fragment_list_categories.xml:
```diff
@@ -1,11 +1,41 @@
+    <androidx.constraintlayout.widget.ConstraintLayout
+        android:layout_width="match_parent"
+        android:layout_height="@dimen/cat_header_height_224dp"
+        android:background="@drawable/bcg_categories"
```
Давай лучше вынесем в отдельный imageView, у него намного больше настроек, щас фото выглядит мыльно. Текущий scaleType никак не работает в ConstraintLayout

aGroupoid commented on app/src/main/res/values/dimens.xml:
```diff
@@ -6,4 +6,9 @@
     <dimen name="text_Size_14">14sp</dimen>
+    <dimen name="text_Size_20">20sp</dimen>
+    <dimen name="cat_header_height_224dp">224dp</dimen>
+    <dimen name="cat_header_text_size">20sp</dimen>
```
Дублируется ресурс

aGroupoid commented on app/src/main/res/values/styles.xml:
```diff
@@ -14,5 +14,14 @@
+        <item name="android:textSize">@dimen/text_Size_14</item>
+        <item name="android:textAllCaps">true</item>
+        <item name="android:textColor">@color/cat_header_txt</item>
+        <item name="android:padding">8dp</item>
```
Вынести в ресурсы

DenisGubernatorov committed:
RA-07-categories-screen
1) Доработка макета карточек
2) Исправление замечаний

aGroupoid commented on app/src/main/res/layout/item_category.xml:
```diff
@@ -0,0 +1,52 @@
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content">
```
Текущая вложенность в ConstraintLayout выглядит лишней, его возможности никак не используются,  это лишняя нагрузка на вёрстку. Почему бы не сделать CardView рутовой View?

aGroupoid commented on app/src/main/res/layout/item_category.xml:
```diff
@@ -0,0 +1,52 @@
+            <ImageView
+                android:layout_width="match_parent"
+                android:layout_height="130dp"
+                android:layout_weight="2"
```
Когда ты пропиcываешь конкретную длину, параметр layout_weight становится бесполезным. Лучше его убрать

aGroupoid commented on app/src/main/res/layout/item_category.xml:
```diff
@@ -0,0 +1,52 @@
+            <TextView
+                style="@style/TitleText14Style"
+                android:layout_width="match_parent"
+                android:layout_height="match_parent"
```
У твоей CardView стоит длина wrap_content, по контенту внутри. А внутри компонентов его ты пишешь match_parent, мы должны посомтреть какая длина у родителя сначала понять, что он wrap_content и сопаставить эту информацию с тем, что здесь на самом деле получается тоже wrap_content. 
Давай не усложнять себе жизнь и напрямую писать в дочерних элементов wrap_content

aGroupoid commented on app/src/main/res/layout/item_category.xml:
```diff
@@ -0,0 +1,52 @@
+                android:layout_height="match_parent"
+                android:layout_marginStart="@dimen/margin_half_8_dp"
+                android:layout_marginTop="@dimen/margin_half_8_dp"
+                android:layout_marginBottom="@dimen/margin_half_8_dp"
```
Есть более униварсальный метод для двух сторон layout_marginVertical. Но в сейчас не хватает marginEnd, длинный текст убдет упираться в границы cardView, на дизайне такого нет. В таком случае все марджины лучше будет просто заменить на margin =

aGroupoid commented on app/src/main/res/layout/item_category.xml:
```diff
@@ -0,0 +1,52 @@
+
+            <ImageView
+                android:layout_width="match_parent"
+                android:layout_height="130dp"
```
Это тоже надо вынести в ресурсы, как и другие элементы до этого

aGroupoid commented on app/src/main/res/layout/fragment_list_categories.xml:
```diff
@@ -1,11 +1,48 @@
+        <ImageView
+            android:layout_width="match_parent"
+            android:layout_height="match_parent"
+            android:scaleType="fitXY"
```
Текущий scaleType как-то сильно растягивает картинку, давай лучше использовать centerCrop

DenisGubernatorov replied:
Честно говоря особой разницы не вижу. Можно заметить только в момент когда меняешь значение, что fitXY расятгивает немного. Есть какие-то инструменты? Или просто "насмотренность"?

DenisGubernatorov committed:
RA-07-categories-screen
1) Доработка макета карточек. Убран лишний констраинт. Для вьюх высота установлена по высоте содержимого. Убраны веса вьюх. Значения для тестовых атрибутов вынесены в ресурсы
2) В заголовке карточки изменен принцип растягивания изображения

aGroupoid reviewed (approved):
Молодец, всё верно!
По поводу инструментов, ты справа снизу можешь потянуть за угол и менять ширину и высоту макета, на широких экранах прям плохо выгляди fitXYZ, а centerCrop хорошо
![image](https://github.com/user-attachments/assets/f250e986-ce28-449a-8934-9ac7b7fa31b5)
